packages/devtools_app/lib/src/screens/memory/primitives/class_name.dart

```diff
@@ -0,0 +1,265 @@
+    assert(
+      !isCore || !isDartOrFlutter,
+      'isCore and isDartOrFlutter must be exclusive',
+    );
+  }
+
+  HeapClassName.fromClassRef(ClassRef? classRef)
+      : library = _library(classRef?.library?.name, classRef?.library?.uri),
+        className = classRef?.name ?? '';
+
+  HeapClassName.fromHeapSnapshotClass(HeapSnapshotClass? theClass)
+      : library =
+            _library(theClass?.libraryName, theClass?.libraryUri.toString()),
+        className = theClass?.name ?? '';
+
+  static String _library(String? libName, String? libUrl) {
+    libName ??= '';
+    if (libName.isNotEmpty) return libName;
+    return libUrl ?? '';
+  }
+
+  final String className;
+  final String library;
+
+  String get fullName => library.isNotEmpty ? '$library/$className' : className;
+
+  bool get isSentinel => className == 'Sentinel' && library.isEmpty;
+
+  /// Detects if a class can retain an object from garbage collection.
+  bool get isWeakEntry {
+    // Classes that hold reference to an object without preventing
+    // its collection.
+    const weakHolders = {
+      '_WeakProperty': 'dart.core',
+      '_WeakReferenceImpl': 'dart.core',
+      'FinalizerEntry': 'dart._internal',
+    };
+
+    if (!weakHolders.containsKey(className)) return false;
+    if (weakHolders[className] == library) return true;
+
+    // If a class lives in unexpected library, this can be because of
+    // (1) name collision or (2) bug in this code.
+    // Throwing exception in debug mode to verify option #2.
+    // TODO(polina-c): create a way for users to add their weak classes
+    // or detect weak references automatically, without hard coding
+    // class names.
+    assert(false, 'Unexpected library for $className: $library.');
+    return false;
+  }
+
+  /// True, if the library is a core library.
+  ///
+  /// I.e. if the library name is empty or does not have prefix
+  /// `dart.` or `package:`.
+  bool get isCore =>
+      library.isEmpty ||
+      (!library.startsWith('dart.') && !library.startsWith('package:'));
+
+  /// True, if the package has prefix `dart.` or has perfix `package:` and is
+  /// published by Dart or Flutter org.
+  bool get isDartOrFlutter {
+    if (library.startsWith('dart.')) return true;
+
+    const packagePrefix = 'package:';
+    if (!library.startsWith(packagePrefix)) return false;
+    final slashIndex = library.indexOf('/');
+    if (slashIndex == -1) return false;
+    final packageName = library.substring(packagePrefix.length, slashIndex);
+
+    return _dartAndFlutterPackages.contains(packageName);
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType) {
+      return false;
+    }
+    return other is HeapClassName && other.fullName == fullName;
+  }
+
+  @override
+  int get hashCode => fullName.hashCode;
+}
+
+/// Packages that are published by dart.dev or flutter.dev.
+const _dartAndFlutterPackages = {
```

Comment:
this will get out of date quickly. Do you have a proposed maintenance plan for this? Additionally, do we really need to get this granular for hiding dart and flutter packages? I think users would expect that "hide dart" hides anything from dart core (dart:package_name excluding dart:ui), and "hide flutter" hides anything from the flutter framework and flutter engine ('package:flutter/...', 'flutter::...', and 'dart:ui')

Reply:
On the tab 'Analysis' we already filter out 'collection' and some other non 'package:flutter/' package.
And Dan confirmed it is how it should be.

My hypothesis is:
1. In memory troubleshooting user will want to focus on non-Google packages, because the Google ones are highly trusted.
2. We do not guarantee to exclude all Google-created package. And it is not a big deal if we miss some of them. User can always add it to the list and file a bug if it bothers them. 
3. We can create a test that pulls html content of the pub.dev pages and verifies all of them are listed. It will be more convenient to find an API to pull the list in a structured way though. But, again, it is nice to have.